src/registers/debug.rs

```diff
@@ -0,0 +1,580 @@
+    ///
+    /// # Safety
+    ///
+    /// The bit representation must be a valid [`Dr7Value`].
+    #[inline]
+    pub const unsafe fn from_bits_unchecked(bits: u64) -> Self {
+        Self { bits }
+    }
+
+    /// Returns the raw value of the fields currently stored.
+    #[inline]
+    pub const fn bits(&self) -> u64 {
+        self.bits
+    }
+
+    /// Returns the [`Dr7Flags`] in this value.
+    #[inline]
+    pub const fn flags(self) -> Dr7Flags {
+        Dr7Flags::from_bits_truncate(self.bits)
+    }
+
+    /// Inserts the specified [`Dr7Flags`] in-place.
+    #[inline]
+    pub fn insert_flags(&mut self, flags: Dr7Flags) {
+        self.bits |= flags.bits();
+    }
+
+    /// Removes the specified [`Dr7Flags`] in-place.
+    #[inline]
+    pub fn remove_flags(&mut self, flags: Dr7Flags) {
+        self.bits &= !flags.bits();
+    }
+
+    /// Toggles the specified [`Dr7Flags`] in-place.
+    #[inline]
+    pub fn toggle_flags(&mut self, flags: Dr7Flags) {
+        self.bits ^= flags.bits();
+    }
+
+    /// Inserts or removes the specified [`Dr7Flags`] depending on the passed value.
+    #[inline]
+    pub fn set_flags(&mut self, flags: Dr7Flags, value: bool) {
+        if value {
+            self.insert_flags(flags);
+        } else {
+            self.remove_flags(flags);
+        }
+    }
+
+    /// Returns the condition field of a debug address register.
+    pub fn get_condition(&self, n: DebugAddressRegisterNumber) -> HwBreakpointCondition {
+        let condition = self.bits.get_bits(HwBreakpointCondition::bit_range(n));
+        HwBreakpointCondition::from_bits(condition).expect("condition should be always valid")
+    }
+
+    /// Sets the condition field of a debug address register.
+    pub fn set_condition(
+        &mut self,
+        n: DebugAddressRegisterNumber,
+        condition: HwBreakpointCondition,
+    ) {
+        self.bits
+            .set_bits(HwBreakpointCondition::bit_range(n), condition as u64);
+    }
+
+    /// Returns the size field of a debug address register.
+    pub fn get_size(&self, n: DebugAddressRegisterNumber) -> HwBreakpointSize {
+        let size = self.bits.get_bits(HwBreakpointSize::bit_range(n));
+        HwBreakpointSize::from_bits(size).expect("condition should be always valid")
+    }
+
+    /// Sets the size field of a debug address register.
+    pub fn set_size(&mut self, n: DebugAddressRegisterNumber, size: HwBreakpointSize) {
+        self.bits
+            .set_bits(HwBreakpointSize::bit_range(n), size as u64);
+    }
+}
+
+/// Debug Control Register (DR7).
+///
+/// Configures debug conditions for debug exceptions.
+#[derive(Debug)]
+pub struct Dr7;
+
+#[cfg(feature = "instructions")]
+mod x86_64 {
+    use super::*;
+    use core::arch::asm;
+
+    impl DebugAddressRegister for Dr0 {
+        #[inline]
+        fn read() -> VirtAddr {
+            let addr;
+
+            unsafe {
+                asm!("mov {}, dr0", out(reg) addr, options(nostack, preserves_flags));
```

Comment:
Here and elsewhere, should this also set "no mem"?

Reply:
Yes, I think so. Done :+1: